clippy_lints/src/str_to_string.rs

```diff
@@ -0,0 +1,57 @@
+use crate::utils::span_lint_and_help;
+
+use if_chain::if_chain;
+use rustc_hir as hir;
+use rustc_hir::*;
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_middle::ty;
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+
+declare_clippy_lint! {
+    /// **What it does:** This lint checks for `.to_string()` method calls on values of type `&str`.
+    ///
+    /// **Why is this bad?** This does a multitude of things to just clone a string. Using `.to_owned()` is more specific.
+    ///
+    /// **Known problems:** None.
+    ///
+    /// **Example:**
+    ///
+    /// ```rust
+    /// // example code where clippy issues a warning
+    /// let _ = "str".to_string()
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// // example code which does not raise clippy warning
+    /// let _ = "str".to_owned()
+    /// ```
+    pub STR_TO_STRING,
+    restriction,
+    "using `to_string()` on a str, which should be `to_owned()`"
+}
+
+declare_lint_pass!(StrToString => [STR_TO_STRING]);
+
+impl LateLintPass<'_> for StrToString {
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &hir::Expr<'_>) {
+        if_chain! {
+            if let ExprKind::MethodCall(path, _, args, _) = &expr.kind;
+            if path.ident.name == sym!(to_string);
+            then {
+                let ty = cx.typeck_results().expr_ty(&args[0]);
+                if let ty::Ref(_, ty, ..) = ty.kind() {
+                    if *ty.kind() == ty::Str {
+                        span_lint_and_help(
+                            cx,
+                            STR_TO_STRING,
+                            expr.span,
+                            "`to_string()` does a multitude of things to just clone a string",
```

Comment:
English is not my mother tongue, so part of my suggestion is maybe good to be ignored :smile: (but keep backticks around `String`)

```suggestion
                            "`to_string()` does a multitude of things than just cloning a `String`",
```